velox/substrait/VeloxToSubstraitType.cpp

```diff
@@ -0,0 +1,139 @@
+      sType->set_allocated_bool_(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_Boolean>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::TINYINT: {
+      sType->set_allocated_i8(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_I8>(&arena));
+      break;
+    }
+    case velox::TypeKind::SMALLINT: {
+      sType->set_allocated_i16(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_I16>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::INTEGER: {
+      sType->set_allocated_i32(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_I32>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::BIGINT: {
+      sType->set_allocated_i64(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_I64>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::REAL: {
+      sType->set_allocated_fp32(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_FP32>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::DOUBLE: {
+      sType->set_allocated_fp64(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_FP64>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::VARCHAR: {
+      sType->set_allocated_varchar(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_VarChar>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::VARBINARY: {
+      sType->set_allocated_binary(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_Binary>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::TIMESTAMP: {
+      sType->set_allocated_timestamp(
+          google::protobuf::Arena::CreateMessage<::substrait::Type_Timestamp>(
+              &arena));
+      break;
+    }
+    case velox::TypeKind::ARRAY: {
+      ::substrait::Type_List* sTList =
+          google::protobuf::Arena::CreateMessage<::substrait::Type_List>(
+              &arena);
+
+      const TypePtr vElementType = vType->asArray().elementType();
+      toSubstraitType(arena, vElementType, sTList->mutable_type());
+      sType->set_allocated_list(sTList);
+
+      break;
+    }
+    case velox::TypeKind::MAP: {
+      ::substrait::Type_Map* sMap =
+          google::protobuf::Arena::CreateMessage<::substrait::Type_Map>(&arena);
+
+      const TypePtr& vMapKeyType = vType->asMap().keyType();
+      const TypePtr& vMapValueType = vType->asMap().valueType();
+
+      toSubstraitType(arena, vMapKeyType, sMap->mutable_key());
+      toSubstraitType(arena, vMapValueType, sMap->mutable_value());
+
+      sType->set_allocated_map(sMap);
+
+      break;
+    }
+    case velox::TypeKind::UNKNOWN:
+    case velox::TypeKind::FUNCTION:
+    case velox::TypeKind::OPAQUE:
+    case velox::TypeKind::INVALID:
+    default:
+      VELOX_UNSUPPORTED("Unsupported type '{}'", std::string(vType->kindName()))
```

Comment:
no need to wrap the output of vType->kindName() in std::string, also, consider using toString() to include more information about the type in the error message.

```
VELOX_UNSUPPORTED("Unsupported type '{}'", vType->toString())
```